builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+
+        Value* const pCmp1 = CreateICmpEQ(pIndex, getIntN(indexBits, 1));
+        pResult = CreateSelect(pCmp1, CreateDPPMov(pIndex, DPPCtrl::DPP_QUAD_PERM1111, 0xf, 0xf, false), pResult);
+
+        Value* const pCmp2 = CreateICmpEQ(pIndex, getIntN(indexBits, 2));
+        pResult = CreateSelect(pCmp2, CreateDPPMov(pIndex, DPPCtrl::DPP_QUAD_PERM2222, 0xf, 0xf, false), pResult);
+
+        Value* const pCmp3 = CreateICmpEQ(pIndex, getIntN(indexBits, 3));
+        pResult = CreateSelect(pCmp3, CreateDPPMov(pIndex, DPPCtrl::DPP_QUAD_PERM3333, 0xf, 0xf, false), pResult);
+    }
+    else
+    {
+        Value* const pCmp0 = CreateICmpEQ(pIndex, getIntN(indexBits, 0));
+        pResult = CreateSelect(pCmp0, CreateDSSwizzle(pIndex, GetDSSwizzleQDMode(0, 0, 0, 0)), pResult);
+
+        Value* const pCmp1 = CreateICmpEQ(pIndex, getIntN(indexBits, 1));
+        pResult = CreateSelect(pCmp1, CreateDSSwizzle(pIndex, GetDSSwizzleQDMode(1, 1, 1, 1)), pResult);
+
+        Value* const pCmp2 = CreateICmpEQ(pIndex, getIntN(indexBits, 2));
+        pResult = CreateSelect(pCmp2, CreateDSSwizzle(pIndex, GetDSSwizzleQDMode(2, 2, 2, 2)), pResult);
+
+        Value* const pCmp3 = CreateICmpEQ(pIndex, getIntN(indexBits, 3));
+        pResult = CreateSelect(pCmp3, CreateDSSwizzle(pIndex, GetDSSwizzleQDMode(3, 3, 3, 3)), pResult);
+    }
+
+    return pResult;
+}
+
+// =====================================================================================================================
+// Create a subgroup quad swap horizontal call.
+Value* BuilderImplSubgroup::CreateSubgroupQuadSwapHorizontal(
+    Value* const pValue,   // [in] The value to swap.
+    const Twine& instName) // [in] Name to give final instruction.
+{
+    if (getContext().GetGfxIpVersion().major >= 8)
+    {
+        return CreateDPPMov(pValue, DPPCtrl::DPP_QUAD_PERM1032, 0xf, 0xf, false);
+    }
+    else
+    {
+        return CreateDSSwizzle(pValue, GetDSSwizzleQDMode(1, 0, 3, 2));
+    }
+}
+
+// =====================================================================================================================
+// Create a subgroup quad swap vertical call.
+Value* BuilderImplSubgroup::CreateSubgroupQuadSwapVertical(
+    Value* const pValue,   // [in] The value to swap.
+    const Twine& instName) // [in] Name to give final instruction.
+{
+    if (getContext().GetGfxIpVersion().major >= 8)
+    {
+        return CreateDPPMov(pValue, DPPCtrl::DPP_QUAD_PERM2301, 0xf, 0xf, false);
+    }
+    else
+    {
+        return CreateDSSwizzle(pValue, GetDSSwizzleQDMode(2, 3, 0, 1));
+    }
+}
+
+// =====================================================================================================================
+// Create a subgroup quadswapdiagonal call.
+Value* BuilderImplSubgroup::CreateSubgroupQuadSwapDiagonal(
+    Value* const pValue,   // [in] The value to swap.
+    const Twine& instName) // [in] Name to give final instruction.
+{
+    if (getContext().GetGfxIpVersion().major >= 8)
```

Comment:
I prefer define a feature list/struct in BuilderImpl and initialize it according gfxip. BuilderImplSubgroup check the feature flag instead of check gfxip every where.
with this policy, it will be easy to enable or disable some features.

Reply:
Good idea!